clang/lib/CodeGen/CGBuiltin.cpp

```diff
@@ -1062,6 +1063,159 @@ CodeGenFunction::emitFlexibleArrayMemberSize(const Expr *E, unsigned Type,
   return Builder.CreateSelect(Cmp, Res, ConstantInt::get(ResType, 0, IsSigned));
 }
 
+namespace {
+
+class ObjectSizeVisitor
+    : public ConstStmtVisitor<ObjectSizeVisitor, const Expr *> {
+  bool SkipASE;
+
+public:
+  ObjectSizeVisitor(bool SkipASE = false) : SkipASE(SkipASE) {}
+
+  const Expr *Visit(const Expr *E) {
+    return ConstStmtVisitor<ObjectSizeVisitor, const Expr *>::Visit(E);
+  }
+
+  const Expr *VisitStmt(const Stmt *S) { return nullptr; }
+
+  const Expr *VisitDeclRefExpr(const DeclRefExpr *E) { return E; }
+  const Expr *VisitMemberExpr(const MemberExpr *E) { return E; }
+  const Expr *VisitArraySubscriptExpr(const ArraySubscriptExpr *E) {
+    return SkipASE ? Visit(E->getBase()) : E;
+  }
+
+  const Expr *VisitCastExpr(const CastExpr *E) {
+    return Visit(E->getSubExpr());
+  }
+  const Expr *VisitParenExpr(const ParenExpr *E) {
+    return Visit(E->getSubExpr());
+  }
+  const Expr *VisitUnaryAddrOf(const clang::UnaryOperator *E) {
+    return Visit(E->getSubExpr());
```

Comment:
For this kind of recursion, it probably makes sense to track whether we're looking at an lvalue or an rvalue.  If "x" is an `int*`, `__bdos(x)` is very different from `__bdos(&x)`.

Reply:
I see what you mean. I kinda want to wave my hands and say that "it's okay, because we're only interested in arrays at this point," but that sounds like Famous Last Words(tm). Do you think that allowing `&` only on an l-value is okay, too restrictive, or not what you had in mind?

Reply:
If you have an array of pointers `int* a[10];`, and someone writes `__bdos(a[b])`, that's equivalent to `void* p = a[b]; __bdos(p)`... and this code can't resolve that kind of construct.  So you need to first look for an "&" (or the equivalent array-to-pointer decay), then look for an array subscript expression inside of that.

If you restrict things enough so each expression you handle is inherently either an lvalue or an rvalue, the result would be logically sound, I guess.  This requires ensuring you don't IgnoreParenImpCasts anywhere because that can look through an lvalue-to-rvalue conversion.  I think it would be more clear with two separate visitors, though.

Reply:
With the latest version of your patch, with the following code, both __builtin_dynamic_object_size() calls fold to 319, which is pretty clearly wrong.  (gcc folds the first call to 312, and the second to -1.)

```
void report(long);
int main(int argc, char **argv) {
char *bar[40];
argc = 1;
report(__builtin_dynamic_object_size(&bar[argc], 1));
report(__builtin_dynamic_object_size(bar[argc], 1));
return 0;
}
```